docs/js-core-3/module-systems/lesson.md

```diff
@@ -0,0 +1,399 @@
+# Module systems in JavaScript
+
+## Learning objectives
+
+- The learner can explain why a module system is useful
+- The learner can use CommonJS to import a module
+- The learner can use relative file paths to import a module
+- The learner can use CommonJS to export from a module
+- ~The learner can use NPM to install all packages in a cloned repo~ I think we would have already covered this
+- The learner can use NPM to install a package in an existing repo
+- The learner can use ECMAScript Modules to import and export a module
+- The learner can identify when to use CJS and when to use ESM
+
+## Why do we need a module system?
+
+### Separating our code into understandable chunks
+
+As our code gets more complex, it becomes more difficult to manage and understand. Real-life programs may have tens of thousands or millions of lines of code. At this scale, it is impossible for people to remember everything about how the code works.
+
+To help with this problem, we can divide our codebase into _modules_, bundling together specific functions and variables. Each of these modules usually has a specific purpose, which allows us to "forget" how some parts of the whole system work, focusing our understanding on a few pieces.
+
+Let's imagine we're working on a codebase that has 1 million lines of code and we need to update the user payments system and how "widgets" in this imaginary system are updated or deleted. This diagram shows how much of the codebase you would need to understand to make those changes with and without a module system:
+
+![No modules code is surrounded by 2 big boxes representing the amount of code needed to understand 2 parts of the system. Modules code is broken up into many separate pieces, so that a much smaller box is needed to represent understanding of the same 2 parts](./assets/separation-of-concerns.png)
+
+As you can see, a module system is very helpful when trying to make changes to a large system.
+
+:::tip
+We sometimes call this _separation of concerns_.
+:::
+
+### Using third-party libraries
+
+Another reason we use a module system is that it allows us to use code written by other programmers. We can be lazy and let others do the work for us!
+
+There are a **huge** number of modules written by other programmers (usually called _packages_) available for us to download and use in our applications. This gives us superpowers to solve complex problems more quickly, since many common problems have already been solved for us. We'll look later at how we can do this in our own applications.
+
+## How does a module system work?
+
+As we've just seen, a module system helps us to split our code into smaller modules.
+
+:::tip
+In nearly every module system, a "module" is really just another name for a file. So we'll just call a "module" a "file" from now on.
+:::
```

Comment:
I always was struck by Bret Victor's bit about how 'file' is quite a poor abstraction. How the 'fundamental units' of computation are expression, line, function, class, module, etc. Wheras a file is just an implementation detail of the underlying hardware.

Not to say this is the best way to teach it (everyone understands what a file is) thought just an interesting aside.